discovery_engine_core/providers/src/lib.rs

```diff
@@ -31,23 +31,180 @@
-pub use client::{Client, CommonQueryParts, HeadlinesQuery, NewsQuery, Query, RankLimit};
 pub use error::Error;
-pub use filter::{Filter, Market};
-pub use newscatcher::{Article as NewscatcherArticle, Response as NewscatcherResponse};
+pub use helpers::{
+    clean_query::clean_query,
+    filter::{Filter, Market},
+    rest_endpoint::RestEndpoint,
+};
 
 mod seal {
-    pub trait Seal {}
+    pub(crate) trait Seal {}
+}
+
+use async_trait::async_trait;
+use url::Url;
+
+/// Abstraction over a provider for a news searching functionality.
+#[async_trait]
+pub trait NewsProvider: Send + Sync {
+    async fn query_news(&self, query: &NewsQuery<'_>) -> Result<Vec<GenericArticle>, Error>;
+}
+
+/// Abstraction over a provider for the latest headlines.
+#[async_trait]
+pub trait HeadlinesProvider: Send + Sync {
+    async fn query_headlines(
+        &self,
+        query: &HeadlinesQuery<'_>,
+    ) -> Result<Vec<GenericArticle>, Error>;
+}
+
+/// Abstraction over a provider for headlines only from trusted sources.
+#[async_trait]
+pub trait TrustedHeadlinesProvider: Send + Sync {
+    async fn query_trusted_sources(
+        &self,
+        query: &TrustedHeadlinesQuery<'_>,
+    ) -> Result<Vec<GenericArticle>, Error>;
+}
+
+#[async_trait]
+pub trait TrendingTopicsProvider: Send + Sync {
+    // TODO: `TrendingTopic` here is the bing specific representation, which we don't really want to expose
+    async fn query_trending_topics(
+        &self,
+        query: &TrendingTopicsQuery<'_>,
+    ) -> Result<Vec<TrendingTopic>, Error>;
+}
+
+pub struct ProviderConfig {
+    pub api_base_url: String,
+    /// Key for accessing the API.
+    pub api_key: String,
+    /// Url path for the news search provider.
+    pub news_provider_path: String,
+    /// Url path for the latest headlines provider.
+    pub headlines_provider_path: String,
+}
+
+pub struct Providers {
+    pub headlines: Arc<dyn HeadlinesProvider>,
+    pub trusted_headlines: Arc<dyn TrustedHeadlinesProvider>,
+    pub news: Arc<dyn NewsProvider>,
+    pub trending_topics: Arc<dyn TrendingTopicsProvider>,
+}
+
+fn create_endpoint_url(str_base_url: &str, path: &str) -> Result<Url, Error> {
+    let mut base_url = Url::parse(str_base_url).map_err(|_| Error::MalformedUrlInConfig {
+        url: str_base_url.into(),
+    })?;
+
+    let mut segments = base_url
+        .path_segments_mut()
+        .map_err(|_| Error::MalformedUrlInConfig {
+            url: str_base_url.into(),
+        })?;
+
+    segments.pop_if_empty();
+    let stripped_path = path.strip_prefix('/').unwrap_or(path);
+    let stripped_path = stripped_path.strip_suffix('/').unwrap_or(stripped_path);
+
+    for new_segment in stripped_path.split('/') {
+        segments.push(new_segment);
+        if new_segment.is_empty() {
+            return Err(Error::MalformedUrlPathInConfig { path: path.into() });
+        }
+    }
+
+    drop(segments);
+    Ok(base_url)
+}
+
+fn select_provider_impl<T: ?Sized>(
```

Comment:
:art: why not simply
```suggestion
fn select_provider<T: ?Sized>(
```
?

Reply:
I think, in the gnews branch from where I copied this, all the classes were called `Impl` at the end, so I guess that's why. But I concur that the shorter name makes sense.